From 4d82b6b6987a727ceaccc20afae8ab554f5de854 Mon Sep 17 00:00:00 2001
From: Suvarna Meenakshi <sumeenak@microsoft.com>
Date: Tue, 26 Jul 2022 22:16:24 +0000
Subject: [PATCH] [caclmgrd]: Allow all incoming traffic on http/https ports
 for restapi Include changes added in PR 3354726 with modification done so
 that multiple threads do not use same config db connector.

Signed-off-by: Suvarna Meenakshi <sumeenak@microsoft.com>
---
 scripts/caclmgrd | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/scripts/caclmgrd b/scripts/caclmgrd
index 4af588e..1d79573 100755
--- a/scripts/caclmgrd
+++ b/scripts/caclmgrd
@@ -206,6 +206,24 @@ class ControlPlaneAclManager(daemon_base.DaemonBase):
         tcp_flags_str = tcp_flags_str[:-1]
         return tcp_flags_str
 
+    def generate_allow_restapi_iptables_commands(self, namespace):
+        # Following ranges from 6 regions are temporarily hardcoded and allowed for Bluebird Service
+        src_ip_range = ["20.44.16.64/27", "13.66.141.96/27", "13.69.229.128/27", "52.162.110.128/27", "52.231.147.224/27", "40.74.146.224/27", "40.78.203.96/27", "20.150.171.224/27", "52.162.110.128/27"]
+
+        dports = ["8081", "8090"]
+
+        allow_restapi_commands = []
+
+        for dport in dports:
+            for src_ip in src_ip_range:
+                ip_ntwrk = ipaddress.ip_network(src_ip, strict=False)
+                if isinstance(ip_ntwrk, ipaddress.IPv4Network):
+                    allow_restapi_commands.append(self.iptables_cmd_ns_prefix[namespace] + "iptables -A INPUT -s {} -p tcp --dport {} -j ACCEPT".format(src_ip, dport))
+                else:
+                    allow_restapi_commands.append(self.iptables_cmd_ns_prefix[namespace] + "ip6tables -A INPUT -s {} -p tcp --dport {} -j ACCEPT".format(src_ip, dport))
+
+        return allow_restapi_commands
+
     def update_feature_present(self):
         feature_tb_info = self.config_db_map[DEFAULT_NAMESPACE].get_table(self.FEATURE_TABLE)
         if feature_tb_info:
@@ -518,6 +536,10 @@ class ControlPlaneAclManager(daemon_base.DaemonBase):
         iptables_cmds.append(self.iptables_cmd_ns_prefix[namespace] + "iptables -A INPUT -p tcp --dport 179 -j ACCEPT")
         iptables_cmds.append(self.iptables_cmd_ns_prefix[namespace] + "ip6tables -A INPUT -p tcp --dport 179 -j ACCEPT")
 
+        # Add iptables/ip6tables commands to allow all restapi/http/https requests
+        if "restapi" in self.feature_present and self.feature_present.get("restapi"):
+            iptables_cmds += self.generate_allow_restapi_iptables_commands(namespace)
+
         # Get current ACL tables and rules from Config DB
         self._tables_db_info = self.config_db_map[namespace].get_table(self.ACL_TABLE)
         self._rules_db_info = self.config_db_map[namespace].get_table(self.ACL_RULE)
-- 
2.17.1

